docs/proposals/20220411-cluster-api-state-metrics.md

```diff
@@ -0,0 +1,359 @@
+
+| Name                      | API Group/Version                       | In-scope |
+|---------------------------|-----------------------------------------|----------|
+| Cluster                   | `cluster.x-k8s.io/v1beta1`              | yes      |
+| ClusterClass              | `cluster.x-k8s.io/v1beta1`              | no       |
+| MachineDeployment         | `cluster.x-k8s.io/v1beta1`              | yes      |
+| MachineSet                | `cluster.x-k8s.io/v1beta1`              | yes      |
+| Machine                   | `cluster.x-k8s.io/v1beta1`              | yes      |
+| KubeadmConfig             | `bootstrap.cluster.x-k8s.io/v1beta1`    | no       |
+| KubeadmConfigTemplate     | `bootstrap.cluster.x-k8s.io/v1beta1`    | no       |
+| KubeadmControlPlane       | `controlplane.cluster.x-k8s.io/v1beta1` | yes      |
+| ClusterResourceSetBinding | `addons.cluster.x-k8s.io/v1beta1`       | no       |
+| ClusterResourceSet        | `addons.cluster.x-k8s.io/v1beta1`       | no       |
+| MachineHealthCheck        | `cluster.x-k8s.io/v1beta1`              | yes      |
+| MachinePool               | `cluster.x-k8s.io/v1beta1`              | yes      |
+
+The relationships between the resources can be found in the [corresponding section in the cluster-api docs].
+
+#### Relationship to kube-state-metrics
+
+There are several CustomResources introduced by Cluster API and Cluster API providers that are similar to core Kubernetes resources.
+Because of that it may make sense to implement metrics inspired by kube-state-metrics.
+
+The following table lists possible mappings from core resources to CRDs:
+
+| kube-state-metrics equivalent | implement for               |
+|-------------------------------|-----------------------------|
+| `Machine`                     | [Pod]                       |
+| `MachineSet`                  | [ReplicaSet]                |
+| `MachineDeployment`           | [Deployment]                |
+| `KubeadmControlPlane`         | [Statefulset], [Deployment] |
+
+CRDs missing in this table:
+
+- `Cluster`
+- `ClusterClass`
+- `ClusterResourceSet`
+- `ClusterResourceSetBinding`
+- `MachineHealthCheck`
+- `MachinePool`
+- `KubeadmConfig`
+- `KubeadmConfigTemplate`
+
+The `Cluster` CR will have important information in their status fields similar to [Pod] metrics like `status.Ready` or `status.Conditions`.
+
+Currently it is not important to implement metrics for `KubeadmConfig` and `KubeadmConfigTemplate` because both only contain configuration data (e.g., passed via cloud-init to the machine). However they may be compared to `ConfigMaps` or `Secrets`.
+
+#### Reusage of kube-state-metrics packages
+
+The proposed exporter should re-use as many packages provided by kube-state-metrics as possible. This allows to re-use flags, configuration and extended functionality like sharding without additional implementation.
```

Comment:
Is this stating will have a vendored dep with kube-state-metrics?

What is an exported? how does kube-state-metrics metrics work?

Can we elaborate the details https://github.com/kubernetes-sigs/cluster-api/pull/6404/files#r849729360?

Reply:
Yes, we will re-use packages and functions from kube-state-metrics to not re-implement it :-)

l:169 scratches the surface of implementation details. I will iterate and improve this to go into more details. 👍

Reply:
I added more detailed paragraphs. Also replaced the usage of the word `exporter` to reduce confusion :-) .

I don't know which kind of diagram would be helpful to visualize unclear details.
A full existing implementation is already in the referenced, to be contributed project: https://github.com/mercedes-benz/cluster-api-state-metrics

I would be happy if you could take another look :-)